### ISSUE TITLE
how to use a Group Managed service account to run a schedule task and make it "Run whether user is logged on or not"?

### ISSUE THREAD

**Issue by dahall** [dahall/TaskScheduler#684]
how can I use a Group Managed service account to run a schedule task and make it "Run whether user is logged on or not"?  
[https://technet.microsoft.com/en-us/library/hh831782.aspx](https://technet.microsoft.com/en-us/library/hh831782.aspx) this is some thing about Group Managed service accounts,   
I tried several ways:  
. use AddTask directly  

        TaskService.Instance.AddTask(&quot;Test&quot;, new MST.DailyTrigger { DaysInterval = 1 },
                                        new MST.ExecAction(&quot;notepad.exe&quot;, &quot;c:\\test.log&quot;, &quot;c:\\program files\\&quot;), &quot;myDomain\\testMSA$&quot;, null, TaskLogonType.None);

only TaskLogonType.None can AddTask successfully, but the task is "Run only when user is logged on"  
  
. RegisterTaskDefinition  

    MyTask.Folder.RegisterTaskDefinition(MyTask.Name, MyTask.Definition,
                                TaskCreation.CreateOrUpdate, &quot;rmad\\testMSA$&quot;, null, TaskLogonType.None);

TaskLogonType.InteractiveToken and TaskLogonType.None can AddTask successfully, but the task is "Run only when user is logged on"  
  
. change Principal  

    MyTask.Definition.Principal.UserId = &quot;rmad\\testMSA$&quot;;
    MyTask.Definition.Principal.LogonType = TaskLogonType.None;
    MyTask.Folder.RegisterTaskDefinition(MyTask.Name, MyTask.Definition);

only TaskLogonType.InteractiveToken can AddTask successfully, but the task is "Run only when user is logged on"  
at the same time, when I assigned TaskLogonType.None and TaskLogonType.Group, I got error "Value does not fall within the expected range"  
  
In short, I can create a schedule task using Group Managed service accounts to run, but just can't make the task "Run whether user is logged on or not"  

Originally posted: 2016-03-22T03:20:11

**Comment by dahall** [comment]
Can you use the system "Task Scheduler" applet to create a task using a Group Managed Service Account? If so, then export the XML and then copy it into your response. I should then be able to tell you how to do it programmatically using the library.  


Originally posted: 2016-03-28T09:29:56

**Comment by dahall** [comment]
Hi dahall, thinks for your reply. I use the following Powershell commands to create a schedule task with GMSA on a windows server 2012R2 machine:  


    $action = New-ScheduledTaskAction  &quot;C:\PowerSlim\test.cmd&quot;
    $trigger = New-ScheduledTaskTrigger -At 20:43 -Daily
    $principal = New-ScheduledTaskPrincipal -UserID er\zz$ -LogonType Password
    Register-ScheduledTask msaTask –Action $action –Trigger $trigger –Principal $principal

then the xml of "msaTask" is like below:  


    &lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-16&quot;?&gt;
    &lt;Task version=&quot;1.3&quot; xmlns=&quot;http://schemas.microsoft.com/windows/2004/02/mit/task&quot;&gt;
      &lt;RegistrationInfo /&gt;
      &lt;Triggers&gt;
        &lt;CalendarTrigger&gt;
          &lt;StartBoundary&gt;2016-03-29T20:43:00&lt;/StartBoundary&gt;
          &lt;Enabled&gt;true&lt;/Enabled&gt;
          &lt;RandomDelay&gt;P0DT0H0M0S&lt;/RandomDelay&gt;
          &lt;ScheduleByDay&gt;
            &lt;DaysInterval&gt;1&lt;/DaysInterval&gt;
          &lt;/ScheduleByDay&gt;
        &lt;/CalendarTrigger&gt;
      &lt;/Triggers&gt;
      &lt;Principals&gt;
        &lt;Principal id=&quot;Author&quot;&gt;
          &lt;UserId&gt;er\zz$&lt;/UserId&gt;
          &lt;LogonType&gt;Password&lt;/LogonType&gt;
          &lt;RunLevel&gt;LeastPrivilege&lt;/RunLevel&gt;
        &lt;/Principal&gt;
      &lt;/Principals&gt;
      &lt;Settings&gt;
        &lt;MultipleInstancesPolicy&gt;IgnoreNew&lt;/MultipleInstancesPolicy&gt;
        &lt;DisallowStartIfOnBatteries&gt;true&lt;/DisallowStartIfOnBatteries&gt;
        &lt;StopIfGoingOnBatteries&gt;true&lt;/StopIfGoingOnBatteries&gt;
        &lt;AllowHardTerminate&gt;true&lt;/AllowHardTerminate&gt;
        &lt;StartWhenAvailable&gt;false&lt;/StartWhenAvailable&gt;
        &lt;RunOnlyIfNetworkAvailable&gt;false&lt;/RunOnlyIfNetworkAvailable&gt;
        &lt;IdleSettings&gt;
          &lt;Duration&gt;PT10M&lt;/Duration&gt;
          &lt;WaitTimeout&gt;PT1H&lt;/WaitTimeout&gt;
          &lt;StopOnIdleEnd&gt;true&lt;/StopOnIdleEnd&gt;
          &lt;RestartOnIdle&gt;false&lt;/RestartOnIdle&gt;
        &lt;/IdleSettings&gt;
        &lt;AllowStartOnDemand&gt;true&lt;/AllowStartOnDemand&gt;
        &lt;Enabled&gt;true&lt;/Enabled&gt;
        &lt;Hidden&gt;false&lt;/Hidden&gt;
        &lt;RunOnlyIfIdle&gt;false&lt;/RunOnlyIfIdle&gt;
        &lt;DisallowStartOnRemoteAppSession&gt;false&lt;/DisallowStartOnRemoteAppSession&gt;
        &lt;UseUnifiedSchedulingEngine&gt;false&lt;/UseUnifiedSchedulingEngine&gt;
        &lt;WakeToRun&gt;false&lt;/WakeToRun&gt;
        &lt;ExecutionTimeLimit&gt;PT72H&lt;/ExecutionTimeLimit&gt;
        &lt;Priority&gt;7&lt;/Priority&gt;
      &lt;/Settings&gt;
      &lt;Actions Context=&quot;Author&quot;&gt;
        &lt;Exec&gt;
          &lt;Command&gt;C:\PowerSlim\test.cmd&lt;/Command&gt;
        &lt;/Exec&gt;
      &lt;/Actions&gt;
    &lt;/Task&gt;



Originally posted: 2016-03-28T23:55:44

**Comment by dahall** [comment]
I thought you needed a trigger that fires at logon?  
  
To accomplish what you have in your PowerShell example:  


    var action = new ExecAction(&quot;C:\PowerSlim\test.cmd&quot;);
    var trigger = new DailyTrigger(DateTime.Today.AddHours(20));
    TaskService.Instance.AddTask(&quot;msaTask&quot;, trigger, action, &quot;er\zz$&quot;, &quot;&quot;, TaskLogonType.Password);

-or-  


    TaskService.Instance.AddTask(&quot;msaTask&quot;, QuickTriggerType.Daily,
       &quot;C:\PowerSlim\test.cmd&quot;, null, &quot;er\zz$&quot;, &quot;&quot;, TaskLogonType.Password);



Originally posted: 2016-03-29T08:50:35

**Comment by dahall** [comment]
these tow ways are not working, here is my test code, I test it on windows server 2012R2:  


    using System;
    using System.Runtime.InteropServices;
    using Microsoft.Win32.TaskScheduler;
    
    namespace WrapperDemo
    {
        internal static class Program
        {
            public static void Main(string[] args)
            {
                try
                {
                    var action = new ExecAction(@&quot;C:\PowerSlim\test.cmd&quot;);
                    var trigger = new DailyTrigger(1);
                    trigger.StartBoundary = DateTime.Today.AddHours(20);
                    TaskService.Instance.AddTask(&quot;msaTask&quot;, trigger, action, @&quot;er\zz$&quot;, &quot;&quot;, TaskLogonType.Password);
                }
                catch (COMException exception)
                {
                    Console.WriteLine(exception.ToString());
                }
    
            }
        }
    }

I got exception like following on the console output:  
  
System.Runtime.InteropServices.COMException (0x8007052E): The username or password is incorrect. (Exception from HRESULT: 0x8007052E)  
   at Microsoft.Win32.TaskScheduler.V2Interop.ITaskFolder.RegisterTaskDefinition(String Path, ITaskDefinition pDefinition, Int32 flags, Object UserId, Object password, TaskLogonType LogonType, Object sddl)  
   at Microsoft.Win32.TaskScheduler.TaskFolder.RegisterTaskDefinition(String Path, TaskDefinition definition, TaskCreation createType, String UserId, String password, TaskLogonType LogonType, String sddl)  
   at Microsoft.Win32.TaskScheduler.TaskService.AddTask(String path, Trigger trigger, Action action, String UserId, String Password, TaskLogonType LogonType)  
   at WrapperDemo.Program.Main(String[] args) in C:\Users\lhuang2\Documents\Visual Studio 2015\Projects\WrapperDemo\WrapperDemo\Program.cs:line 20  


Originally posted: 2016-03-30T00:43:01

**Comment by dahall** [comment]
This discussion has been copied to a work item. Click [here](https://taskscheduler.codeplex.com/workitem/12362) to go to the work item and continue the discussion.  


Originally posted: 2016-03-30T08:13:28

**Comment by dahall** [comment]
Please check the link to the created issue (from previous post). I have posted some code and test assemblies for you to test.  


Originally posted: 2016-04-02T12:39:57

**Comment by dahall** [comment]
Hi dahall, I tried, your assemblies work well.  


Originally posted: 2016-04-05T20:21:32

**Comment by dahall** [comment]
Thanks. I'll get this out shortly as 2.5.17.  


Originally posted: 2016-04-07T07:39:24
